### PR TITLE
Add checking stake approval before requesting approval to skip approval

### PR DESCRIPTION
**Pull request by Aerilym**
When registering a session node, the operator must approve the rewards contract to take the required stake amount, this only needs to be done once. This change adds a check to the registration stages to see if the operator as already approved the required stake amount.

### CODE REVIEW

File: packages/contracts/hooks/SENT.tsx [hunk 1]
```diff
@@ -53,9 +59,54 @@ export function useSENTBalanceQuery(
   };
 }
 
+export type SENTAllowanceQuery = ReadContractQuery & {
+  /** Get the session token allowance */
+  getAllowance: () => void;
+  /** The session token allowance for a contract */
+  allowance: ReadContractData<typeof SENTAbi, 'allowance', [Address, Address]>;
```
Comment (yougotwill) [inline packages/contracts/hooks/SENT.tsx:66]: How is the contract read here separate to the getAllowance function?

File: packages/contracts/hooks/SENT.tsx [hunk 2]
```diff
@@ -66,16 +117,66 @@ export function useProxyApproval({
...
 }): UseProxyApprovalReturn {
+  const [hasEnoughAllowance, setHasEnoughAllowance] = useState<boolean>(false);
+  const { address } = useAccount();
+  const {
+    allowance,
+    getAllowance,
+    status: readStatus,
+  } = useAllowanceQuery({
+    contractAddress,
+    chainId: chains[CHAIN.TESTNET].id,
```
Comment (yougotwill) [inline packages/contracts/hooks/SENT.tsx:128]: Should we not start having a env flag or global variable for the chain we want to use? That way down the line it's a one line switch from testnet to mainet?

File: packages/contracts/hooks/SENT.tsx [hunk 3]
```diff
@@ -66,16 +117,66 @@ export function useProxyApproval({
...
+  };
+
+  const approveWrite = () => {
+    if (readStatus !== CONTRACT_READ_STATUS.SUCCESS) {
+      throw new Error('Checking if current allowance is sufficient');
+    }
+
+    if (allowance >= tokenAmount) {
+      setHasEnoughAllowance(true);
+      if (!isProduction()) {
```
Comment (yougotwill) [inline packages/contracts/hooks/SENT.tsx:148]: If production it's okay to allow this?

File: packages/contracts/hooks/SENT.tsx [hunk 4]
```diff
@@ -66,16 +117,66 @@ export function useProxyApproval({
...
       args: [contractAddress, tokenAmount],
     });
   };
 
-  return { approve, writeStatus, error };
+  const status = useMemo(() => {
+    if (readStatus === CONTRACT_READ_STATUS.SUCCESS && hasEnoughAllowance) {
+      return 'success';
+    }
+
```
Comment (yougotwill) [inline packages/contracts/hooks/SENT.tsx:164]: I prefer new lines between if statement blocks. You can do one or the other please 🤷

File: packages/contracts/hooks/contract-hooks.tsx [hunk 5]
```diff
@@ -203,6 +209,8 @@ export type ContractReadQueryFetchOptions<Args = unknown> = {
   startEnabled?: boolean;
   /** The initial arguments to use when fetching the contract. */
   args?: Args;
+  /** Chain id the contract is on */
+  chainId: number;
```
Comment (yougotwill) [inline packages/contracts/hooks/contract-hooks.tsx:213]: Can we not inherit from `ContractWriteQueryFetchOptions` here?

**Review by yougotwill — CHANGES_REQUESTED** [review]
left some comments

File: packages/contracts/hooks/contract-hooks.tsx (hunk 5 above)
Comment (Aerilym) [reply packages/contracts/hooks/contract-hooks.tsx:213]: I dont want to pollute the read query options as we'll likely add more props to the write query than just chainId, so we want them to be separate as they're unrelated. I have plans to split this file into two when i get time

File: packages/contracts/hooks/SENT.tsx (hunk 1 above)
Comment (Aerilym) [reply packages/contracts/hooks/SENT.tsx:66]: I'm not sure i follow. `getAllowance()` triggers the getting of the allowance (the query), and allowance is the allowance value (which changes based on the query)

File: packages/contracts/hooks/SENT.tsx (hunk 2 above)
Comment (Aerilym) [reply packages/contracts/hooks/SENT.tsx:128]: We're going to build a network management system as we need to pass this through depending on the network and so the chain you're on. You can switch chain at any time. its hard coded here as we dont have that built yet

File: packages/contracts/hooks/SENT.tsx (hunk 3 above)
Comment (Aerilym) [reply packages/contracts/hooks/SENT.tsx:148]: if not production :)

File: packages/contracts/hooks/SENT.tsx (hunk 4 above)
Comment (Aerilym) [reply packages/contracts/hooks/SENT.tsx:164]: Ah this was missed. fixed

File: packages/contracts/hooks/SENT.tsx (hunk 1 above)
Comment (Aerilym) [reply packages/contracts/hooks/SENT.tsx:66]: So this is the query convention for tanstack query :)

**Pull request merged**